### ISSUE TITLE
Linear Regression Model Returning NaN: 0.0 Values Being Discarded in MutableDataset

### ISSUE THREAD

**Issue by rithin-pullela-aws** [oracle/tribuo#385]
**Describe the bug**

I’m working on a linear regression task using Tribuo, and the model is returning NaN values during prediction. 

After debugging, I discovered that the model’s weights were NaN, and I believe the issue is that the MutableDataset<Regressor> is discarding feature entries where the feature value is 0.0. This behavior seems to stem from [here](https://github.com/oracle/tribuo/blob/81605a2317cbb81095f474022a9e0307880d8001/Core/src/main/java/org/tribuo/CategoricalInfo.java#L187). The count of features (in MutableDataset -> featureMap -> m) does not include entries with a value of 0.0, leading to inconsistent feature counts and likely causing issues during model training (Refer to Screenshot1 below).

**To Reproduce**

Steps to reproduce the behaviour:
1. Set up a linear regression task in Tribuo, using a MutableDataset<Regressor>.
2. Include feature entries with 0.0 values in your dataset.
3. Observe the MutableDataset has missing feature counts.

- Data set I used: https://gist.github.com/wrigleyDan/a83a5d8294aa0ed493e4feb8cc9d7433
- predict_data.csv: I used few rows from train dataset
- Code Snippet:
```
        var regressionFactory = new RegressionFactory();
        var csvLoader = new CSVLoader<>(',',regressionFactory);
        DataSource<Regressor> features = csvLoader.loadDataSource(Paths.get("features_5000_queries.csv"),"neuralness");
        MutableDataset<Regressor> trainData = new MutableDataset<>(features);


        Double learningRate = 0.01;
        Double momentumFactor = 0.0;
        SGD.Momentum momentum = SGD.Momentum.STANDARD;
        StochasticGradientOptimiser optimiser = SGD.getSimpleSGD(learningRate, momentumFactor, momentum);
        RegressionObjective objective = new SquaredLoss();
        LinearSGDTrainer linearSGDTrainer  = new LinearSGDTrainer(objective, optimiser, 1000, -1, 1, 12345L);
        Model<Regressor> regressionModel = linearSGDTrainer.train(trainData);

        DataSource<Regressor> preds = csvLoader.loadDataSource(Paths.get("predict_data.csv"), "neuralness");
        MutableDataset<Regressor> testData = new MutableDataset<>(preds);

        List<Prediction<Regressor>> p =regressionModel.predict(testData);
```

**Expected behaviour**

1. The linear regression model should handle all data points, including those with feature values of 0.0, and produce reasonable predictions.
2. The model weights should not be NaN.
3. The Predictions should not be NaN.

**Additional Questions**

1. Why are 0.0 values ignored in the [observe function](https://github.com/oracle/tribuo/blob/81605a2317cbb81095f474022a9e0307880d8001/Core/src/main/java/org/tribuo/CategoricalInfo.java#L187) ?
2. Is this the expected behavior in Tribuo, or should these values(0.0) be included in the training data?
3. Is there a recommended way to handle features with a value of 0.0 in Tribuo?
4. Would normalizing the data (e.g., to a range like 1-2) help avoid this issue?

**Screenshots**
Screenshot1 :
<img width="1283" alt="Tribuo_1" src="https://github.com/user-attachments/assets/af3b6215-0b50-4cbb-8e07-0caf44abfb22" />

**System information:**

 - Tribuo Version: 4.2.1
 - OS: macOS
 - Java Version: openjdk 21.0.5 

**Additional context**

The dataset I provided contains 5,000 examples, and I've noticed that the feature count is consistently 5,000 minus the number of zero values. For instance, in case of `f_5_num_results`, it has 760 Zeros and the count in the map can be found as 4240(5000-760). This behavior is illustrated in screenshot1.

**Comment by Craigacp** [comment]
The model weights became NaN because SGD with a constant learning rate and no learning rate scheduler isn't a great optimizer, so the model weights diverged to negative infinity and then got a NaN loss which Tribuo helpfully backpropagated into the weights. If you use `AdaGrad` or tune the learning rate then it doesn't diverge (in my tests anyway) and you get a model which has real valued weights and doesn't make NaN predictions. Note, I've not actually tested if the model it produces is good or not, I'm not sure what's an appropriate test set for this data.

The decision to treat zero values specially is intentional and comes from one of Tribuo's initial design requirements to handle mid 2010's NLP systems with large sparse feature vectors with many zero values. All examples in Tribuo are implicitly sparse, we don't represent zero values anywhere in the examples and they are tracked through a separate mechanism in the feature maps. As the weight times the zero in a linear regression will be zero it doesn't need to be computed, and we have similar behaviour for all our other `Trainer` implementations. If the trainer requires dense features (e.g. TensorFlow) then we build a [dense representation before training](https://github.com/oracle/tribuo/blob/main/Interop/Tensorflow/src/main/java/org/tribuo/interop/tensorflow/DenseFeatureConverter.java#L125). Underneath the covers Tribuo will decide if any given vector should be dense or sparse based on if it would compute faster, though we currently mostly default to sparse. In the next release it'll be a bit more aggressive about making things dense as it improves performance, and also with images & text embeddings there are a lot more dense feature spaces around these days than we had for our production systems when this design choice was made.
 
This is the code I ran:
```java
RegressionFactory regressionFactory = new RegressionFactory();
double learningRate = 0.01;
var optimiser = new AdaGrad(learningRate);
RegressionObjective objective = new SquaredLoss();
LinearSGDTrainer linearSGDTrainer = new LinearSGDTrainer(objective, optimiser, 1000, -1, 1, 12345L);

var regexMappingProcessors = new HashMap<String,FieldProcessor>();
regexMappingProcessors.put("f_.*", new DoubleFieldProcessor("f_.*"));

var responseProcessor = new FieldResponseProcessor<>("neuralness","-1",regressionFactory);
var metadataExtractors = new ArrayList<FieldExtractor<?>>();
metadataExtractors.add(new IdentityExtractor("query_string"));
metadataExtractors.add(new IdentityExtractor("ndcg"));

var rowProcessor = new RowProcessor.Builder<Regressor>()
        .setMetadataExtractors(metadataExtractors)
        .setRegexMappingProcessors(regexMappingProcessors)
        .build(responseProcessor);

var csvSource = new CSVDataSource<Regressor>(Paths.get("<path-to-file>/features.csv"),rowProcessor,true);
MutableDataset<Regressor> trainData = new MutableDataset<>(csvSource);

LinearSGDModel regressionModel = linearSGDTrainer.train(trainData);

System.out.println(regressionModel.getWeightsCopy());
```

and I get the output:
```
DenseMatrix(dim1=1,dim2=10,values=
	row 0 [ 0.035595772230248,-0.002271644203614,-0.034022036223791, 0.020566662867170, 0.000006491115183, 0.073114582824161,-0.041301633274970, 0.156873656590218, 0.226363813694671, 0.381500037900483];
)
```

**Comment by rithin-pullela-aws** [comment]
Thank you so much for your detailed explanation! I really appreciate the insight into Tribuo's design choices and how it handles zero values. That definitely clarifies a lot about the behavior I was seeing with the feature map and the NaN values during training.

I have experimented with Ada Grad and it resulted in better results.

**Comment by rithin-pullela-aws** [comment]
Closing the bug, since having 0 values do not impact the training.
